experiments/kompanion/cmd/summary/summary.go

```diff
@@ -0,0 +1,398 @@
+	gkMap[status] = count + 1
+
+	if r.totals == nil {
+		r.totals = make(map[string]map[string]int)
+	}
+	gkMap, present = r.totals[gk]
+	if !present {
+		gkMap = make(map[string]int)
+		r.totals[gk] = gkMap
+	}
+	count = gkMap[status]
+	// Can ignore not present because count will default to 0.
+	gkMap[status] = count + 1
+}
+
+func (r *Result) Print() {
+	r.countLock.Lock()
+	defer r.countLock.Unlock()
+	for ns, nsMap := range r.counts {
+		log.Printf("Namespace: %s\n", ns)
+		for gk, gkMap := range nsMap {
+			first := true
+			info := "("
+			for status, count := range gkMap {
+				if first {
+					info += fmt.Sprintf("%s: %d", status, count)
+					first = false
+				} else {
+					info += fmt.Sprintf(", %s: %d", status, count)
+				}
+			}
+			info += ")"
+			log.Printf("- GroupKind: %s %s\n", gk, info)
+		}
+	}
+	log.Printf("Totals: \n")
+	for gk, gkMap := range r.totals {
+		first := true
+		info := "("
+		for status, count := range gkMap {
+			if first {
+				info += fmt.Sprintf("%s: %d", status, count)
+				first = false
+			} else {
+				info += fmt.Sprintf(", %s: %d", status, count)
+			}
+		}
+		info += ")"
+		log.Printf("- GroupKind: %s %s\n", gk, info)
+	}
+}
+
+// Task is implemented by our namespace-collection routine, or anything else we want to run in parallel.
+type Task interface {
+	Run(ctx context.Context) error
+}
+
+// tracks the namespaces to be exported.
+// thread safe.
+type taskQueue struct {
+	mu    sync.Mutex
+	tasks []Task // will be treated as a FIFO queue
+}
+
+func (n *taskQueue) GetWork() Task {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if len(n.tasks) == 0 {
+		return nil
+	}
+
+	workItem := n.tasks[0]
+	n.tasks = n.tasks[1:]
+
+	return workItem
+}
+
+func (n *taskQueue) AddTask(t Task) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	n.tasks = append(n.tasks, t)
+}
+
+type dumpResourcesTask struct {
+	// Result accumulates the overall summary
+	Summary *Result
+
+	// Namespace is the namespace to filter down
+	Namespace string
+
+	DynamicClient *dynamic.DynamicClient
+
+	// Resources is the list of resources to query
+	Resources []schema.GroupVersionResource
+}
+
+func (t *dumpResourcesTask) Run(ctx context.Context) error {
+	// klog := klog.FromContext(ctx)
+
+	for _, gvr := range t.Resources {
+		var resources *unstructured.UnstructuredList
+		if t.Namespace != "" {
+			r, err := t.DynamicClient.Resource(gvr).Namespace(t.Namespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("fetching gvr %s resources in namespace %s: %w", gvr, t.Namespace, err)
+			}
+			resources = r
+
+		} else {
+			r, err := t.DynamicClient.Resource(gvr).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("fetching gvr %s resources: %w", gvr, err)
+			}
+			resources = r
+		}
+
+		for _, r := range resources.Items {
+			status, err := getStatus(r, gvr)
+			if err != nil {
+				log.Printf("Got error retrieving status of type gvr %s, %v", gvr, err)
+			}
+			gk := gvr.Resource + "." + gvr.Group
```

Comment:
wait I am confused -- is this actually the group kind then ?

Reply:
Group and Kind should uniquely identifies a type of resource. Version identifies which (of several) access mechanisms you want to use to get that resource. If we include the gvr we will can end up multi-counting the resource. We should probably actually add code in utils.GetResources() to defend against that. I don't believe we are seeing instances of that today because we generally remove the v1alphaX when we add the v1betaY. Also we don't currently have any V1s. However we probably should add some protections to GetResources().